sotodlib/site_pipeline/update_preprocess_plots.py

```diff
@@ -0,0 +1,256 @@
+import os
+import yaml
+import time
+import logging
+import numpy as np
+import argparse
+import traceback
+from typing import Optional
+import copy
+
+from sotodlib import core
+import sotodlib.site_pipeline.util as sp_util
+from sotodlib.preprocess import _Preprocess, Pipeline, processes
+
+logger = sp_util.init_logger("preprocess")
+
+def _get_preprocess_context(configs, context=None):
+    if type(configs) == str:
+        configs = yaml.safe_load(open(configs, "r"))
+    
+    if context is None:
+        context = core.Context(configs["context_file"])
+        
+    if type(context) == str:
+        context = core.Context(context)
+    
+    # if context doesn't have the preprocess archive it in add it
+    # allows us to use same context before and after calculations
+    found=False
+    if context.get("metadata") is None:
+        context["metadata"] = []
+
+    for key in context.get("metadata"):
+        if key.get("unpack") == "preprocess":
+            found=True
+            break
+    if not found:
+        context["metadata"].append( 
+            {
+                "db" : configs["archive"]["index"],
+                "unpack" : "preprocess"
+            }
+        )
+    return configs, context
+
+def _get_groups(obs_id, configs, context):
+    group_by = np.atleast_1d(configs['subobs'].get('use', 'detset'))
+    for i, gb in enumerate(group_by):
+        if gb.startswith('dets:'):
+            group_by[i] = gb.split(':',1)[1]
+
+        if (gb == 'detset') and (len(group_by) == 1):
+            groups = context.obsfiledb.get_detsets(obs_id)
+            return group_by, [[g] for g in groups]
+        
+    det_info = context.get_det_info(obs_id)
+    rs = det_info.subset(keys=group_by).distinct()
+    groups = [[b for a,b in r.items()] for r in rs]
+    return group_by, groups
```

Comment:
We're having a lot of repeated code between here, `preprocess_tod` and `preprocess_obs` we need to find places where code is repeated, make a common preprocess utility module to import those functions from, and then modify the existing scripts to use those utilities. If you don't do it in this PR please create an issue for it.

Reply:
I've added these functions to `util.py` and `update_preprocess_plots` imports them. I will leave the change for `preprocess_tod` and `preprocess_obs` for another PR.